### PR TITLE
feat(masthead): overflow scrolling animation

### PR DESCRIPTION
**Pull request by ariellalgilmore**
### Related Ticket(s)

#4164 

### Description

create scrolling animation for overflow
ezgif com-gif-maker

### Changelog

**New**

- additional styles
- use intersectionObserver to check for turning on and off left and right carets

### CODE REVIEW

File: packages/react/src/components/Masthead/HeaderNavContainer.js [hunk 1]
```diff
@@ -18,97 +17,80 @@ const { prefix } = settings;
...
-        })
-      );
-    }
-  }, []);
+    setIO(
+      new IntersectionObserver(
+        records => {
+          records.forEach(record => {
+            if (record.target.classList.contains('sub-content-left')) {
+              document.querySelector('.bx--header__nav-caret-left').hidden =
```
Comment (asudoh) [inline packages/react/src/components/Masthead/HeaderNavContainer.js:46]: Should be grabbed from React `ref`.

File: packages/react/src/components/Masthead/HeaderNavContainer.js [hunk 2]
```diff
@@ -18,97 +17,80 @@ const { prefix } = settings;
...
+                record.isIntersecting;
+            }
+            if (record.target.classList.contains('sub-content-right')) {
+              document.querySelector('.bx--header__nav-caret-right').hidden =
```
Comment (asudoh) [inline packages/react/src/components/Masthead/HeaderNavContainer.js:50]: Should be grabbed from React `ref`.

File: packages/react/src/components/Masthead/HeaderNavContainer.js [hunk 3]
```diff
@@ -18,97 +17,80 @@ const { prefix } = settings;
...
-          totalNavWidth - 80
-      ) {
-        setShowRightCaret(true);
-      }
-      if (headerNavContainerRef.current.scrollLeft > 0) {
-        setShowLeftCaret(true);
-      }
+    if (io) {
+      io.observe(document.querySelector('.sub-content-left'));
+      io.observe(document.querySelector('.sub-content-right'));
```
Comment (asudoh) [inline packages/react/src/components/Masthead/HeaderNavContainer.js:66]: Should be grabbed from React `ref`.

File: packages/styles/scss/components/masthead/_masthead.scss [hunk 4]
```diff
@@ -307,6 +314,27 @@ $search-transition-timing: 95ms;
...
+  }
+
+  .sub-content-right {
+    position: absolute;
+    top: 0;
+    right: 1px;
+    width: 1px;
+    height: 100%;
+    //background-color: purple;
+    //clip: rect(0,0,0,0);
```
Comment (asudoh) [inline packages/styles/scss/components/masthead/_masthead.scss:335]: Did you mean to remove these lines?

File: packages/styles/scss/components/masthead/_masthead.scss [hunk 5]
```diff
@@ -307,6 +314,27 @@ $search-transition-timing: 95ms;
...
 
+  .sub-content-left {
+    position: absolute;
+    top: 0;
+    left: 0;
+    width: 1px;
+    height: 100%;
+    //background-color: purple;
+
+    //clip: rect(0,0,0,0);
```
Comment (asudoh) [inline packages/styles/scss/components/masthead/_masthead.scss:325]: Did you mean to remove these lines?

**Review by asudoh — COMMENTED** [review]
Great to see this seems ready-to-review @ariellalgilmore! 🎉 BTW did you mean to make this PR out of draft?

**Comment by ariellalgilmore** [comment]
Thanks @asudoh!! Updated with refs and deleted comments from css! I left it as a draft because one thing I still was trying to research is why sometimes the right click needs to be pressed twice. Please let me know if you have any insights on this. Thanks!

**Comment by asudoh** [comment]
Looks that the button visibility has adverse effect on container/content intersection state:

intersection-caret-0
intersection-caret-1

File: packages/styles/scss/components/masthead/_masthead.scss [hunk 6]
```diff
@@ -308,13 +316,30 @@ $search-transition-timing: 95ms;
...
+    height: 100%;
+  }
+
+  .sub-content-right {
+    position: absolute;
+    top: 0;
+    right: 0;
+    width: 1px;
+    height: 100%;
+  }
```
Comment (asudoh) [inline packages/styles/scss/components/masthead/_masthead.scss:333]: Do we want to add `.bx--` prefix? If not, probably `.dds--` prefix?

**Review by asudoh — APPROVED** [review]
LGTM 👍 just one thing - Thanks @ariellalgilmore! 🎉

File: packages/styles/scss/components/masthead/_masthead.scss (hunk 6 above)
Comment (ariellalgilmore) [reply packages/styles/scss/components/masthead/_masthead.scss:333]: yes agree! Good catch!

Comment (ariellalgilmore) [reply packages/styles/scss/components/masthead/_masthead.scss]: just updated! Thanks!

File: packages/styles/scss/components/masthead/_masthead.scss [hunk 7]
```diff
@@ -308,13 +316,30 @@ $search-transition-timing: 95ms;
...
 
+  .#{$prefix}-sub-content-left {
+    position: absolute;
+    top: 0;
+    left: 0;
+    width: 1px;
+    height: 100%;
+  }
+
+  .#{$prefix}-sub-content-right {
```
Comment (asudoh) [inline packages/styles/scss/components/masthead/_masthead.scss:327]:
```suggestion
  .#{$prefix}--sub-content-right {
```

File: packages/styles/scss/components/masthead/_masthead.scss [hunk 8]
```diff
@@ -308,13 +316,30 @@ $search-transition-timing: 95ms;
     }
   }
 
+  .#{$prefix}-sub-content-left {
```
Comment (asudoh) [inline packages/styles/scss/components/masthead/_masthead.scss:319]:
```suggestion
  .#{$prefix}--sub-content-left {
```

File: packages/react/src/components/Masthead/HeaderNavContainer.js [hunk 9]
```diff
@@ -18,100 +17,96 @@ const { prefix } = settings;
...
-        onClick={paginateLeft}>
-        <CaretLeft20 />
-      </button>
-      <div
-        className={`${prefix}--header__nav-container`}
-        ref={headerNavContainerRef}>
-        {children}
+      <div className={`${prefix}--header__nav-container`} ref={containerRef}>
+        <div className={`${prefix}--header__nav-content`} ref={contentRef}>
+          <div className={`${prefix}-sub-content-left`} ref={contentLeftRef} />
```
Comment (asudoh) [inline packages/react/src/components/Masthead/HeaderNavContainer.js:88]:
```suggestion
          <div className={`${prefix}--sub-content-left`} ref={contentLeftRef} />
```

File: packages/react/src/components/Masthead/HeaderNavContainer.js [hunk 10]
```diff
@@ -18,100 +17,96 @@ const { prefix } = settings;
...
+          <div
+            className={`${prefix}-sub-content-right`}
```
Comment (asudoh) [inline packages/react/src/components/Masthead/HeaderNavContainer.js:90]:
```suggestion
            className={`${prefix}--sub-content-right`}
```

**Review by asudoh — COMMENTED** [review]
To follow the BEM naming structure - Thanks @ariellalgilmore!

**Review by jeffchew — CHANGES_REQUESTED** [review]
Hey @ariellalgilmore , not sure if this is fully working, I can only click the arrow once and looks like there's still more items not fully visible.

overflow

**Comment by ariellalgilmore** [comment]
closing due to conflicts. Will re-open new PR

**Pull request closed without merging**